File: lib/importio-client-libs-master/csharp/4.0/ImportIO.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net;
using System.IO;
using System.Web;
using System.Net;

// Download the Newtonsoft JSON library here http://james.newtonking.com/projects/json-net.aspx
using Newtonsoft.Json;
using System.Threading;
using System.Collections.Concurrent;

namespace MinimalCometLibrary
{
    delegate void QueryHandler(Query query, Dictionary<String, Object> data);

    class Query
    {
        int jobsCompleted = 0;
        int jobsStarted = 0;
        int jobsSpawned = 0;
        private bool finished = false;

        public bool isFinished { get { return finished; } set { finished = value; } }
        public QueryHandler queryCallback;

        Dictionary<String, Object> queryInput;

        public Query(Dictionary<String, Object> queryInput, QueryHandler queryCallback)
        {
            this.queryInput = queryInput;
            this.queryCallback = queryCallback;
        }

        public void OnMessage(Dictionary<String, Object> data)
        {
            String messageType = (String)data["type"];

            Console.WriteLine((String)data["type"]);

            switch (messageType)
            {
                case "SPAWN":
                    jobsSpawned++;
                    break;
                case "INIT":
                case "START":
                    jobsStarted++;
                    break;
                case "STOP":
                    jobsCompleted++;
                    break;
            }

            finished = jobsStarted == jobsCompleted && jobsSpawned + 1 == jobsStarted && jobsStarted > 0;

            if (messageType.Equals("ERROR") || messageType.Equals("UNAUTH") || messageType.Equals("CANCEL"))
            {
                finished = true;
            }

            queryCallback(this, data);
        }

    }

    class ImportIO
    {
        private String host { get; set; }
        private int port { get; set; }

        private Guid userGuid;
        private String apiKey;

        private static String messagingChannel = "/messaging";
        private String url;

        private int msgId = 0;
        private String clientId;

        private Boolean isConnected;

        CookieContainer cookieContainer = new CookieContainer();

        Dictionary<Guid, Query> queries = new Dictionary<Guid, Query>();

        private BlockingCollection<Dictionary<String, Object>> messageQueue = new BlockingCollection<Dictionary<string,object>>();

        public ImportIO(String host = "http://query.import.io", Guid userGuid = default(Guid), String apiKey = null)
        {
            this.userGuid = userGuid;
            this.apiKey = apiKey;

            this.url = host + "/query/comet/";
            clientId = null;
        }

        public void Login(String username, String password, String host = "http://api.import.io")
        {
            Console.WriteLine("Logging in");
            String loginParams = "username=" + HttpUtility.UrlEncode(username) + "&password=" + HttpUtility.UrlEncode(password);
            String searchUrl = host + "/auth/login";
            HttpWebRequest loginRequest = (HttpWebRequest)WebRequest.Create(searchUrl);

            loginRequest.Method = "POST";
            loginRequest.ContentType = "application/x-www-form-urlencoded";
            loginRequest.ContentLength = loginParams.Length;

            loginRequest.CookieContainer = cookieContainer;

            using (Stream dataStream = loginRequest.GetRequestStream())
            {
                dataStream.Write(System.Text.UTF8Encoding.UTF8.GetBytes(loginParams), 0, loginParams.Length);

                HttpWebResponse loginResponse = (HttpWebResponse)loginRequest.GetResponse();


                if (loginResponse.StatusCode != HttpStatusCode.OK)
                {
                    throw new Exception("Could not log in, code:" + loginResponse.StatusCode);
                }
                else
                {
                    foreach (Cookie cookie in loginResponse.Cookies)
                    {
                        if (cookie.Name.Equals("AUTH"))
                        {
                            // Login was successful
                            Console.WriteLine("Login Successful");
                        }
                    }

                }
            }
        }

        public List<Dictionary<String, Object>> Request(String channel, Dictionary<String, Object> data = null, String path = "", Boolean doThrow = true)
        {
            Dictionary<String, Object> dataPacket = new Dictionary<String, Object>();
            dataPacket.Add("channel", channel);
            dataPacket.Add("connectionType", "long-polling");
            dataPacket.Add("id", (msgId++).ToString());

            if (this.clientId != null)
                dataPacket.Add("clientId", this.clientId);

            if (data != null)
            {
                foreach (KeyValuePair<String, Object> entry in data)
                {
                    dataPacket.Add(entry.Key, entry.Value);
                }
            }

            String url = this.url + path;

            if (apiKey != null)
            {
                url += "?_user=" + HttpUtility.UrlEncode(userGuid.ToString()) + "&_apikey=" + HttpUtility.UrlEncode(apiKey);
            }

            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
            request.AutomaticDecompression = DecompressionMethods.GZip;
            request.Method = "POST";
            request.ContentType = "application/json;charset=UTF-8";
            request.Headers.Add(HttpRequestHeader.AcceptEncoding, "gzip");
            String dataJson = JsonConvert.SerializeObject(new List<Object>() { dataPacket });

            request.ContentLength = dataJson.Length;

            request.CookieContainer = cookieContainer;

            using (Stream dataStream = request.GetRequestStream())
            {
                dataStream.Write(System.Text.UTF8Encoding.UTF8.GetBytes(dataJson), 0, dataJson.Length);
                try
                {
                    HttpWebResponse response = (HttpWebResponse)request.GetResponse();

                    using (StreamReader responseStream = new StreamReader(response.GetResponseStream()))
                    {
                        String responseJson = responseStream.ReadToEnd();
                        List<Dictionary<String, Object>> responseList = JsonConvert.DeserializeObject<List<Dictionary<String, Object>>>(responseJson);
                        foreach (Dictionary<String, Object> responseDict in responseList)
                        {
                            if (responseDict.ContainsKey("successful") && (bool)responseDict["successful"] != true)
                            {
                                if (doThrow)
                                    throw new Exception("Unsucessful request");
                            }

                            if (!responseDict["channel"].Equals(messagingChannel)) continue;

                            if (responseDict.ContainsKey("data"))
                            {
                                messageQueue.Add(((Newtonsoft.Json.Linq.JObject)responseDict["data"]).ToObject<Dictionary<String, Object>>());
                            }

                        }

                        return responseList;
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine("Error occurred {0}", e.Message);
                    return new List<Dictionary<String, Object>>();
                }
                
            }
        }

        public void Handshake()
        {
            Dictionary<String, Object> handshakeData = new Dictionary<String, Object>();
            handshakeData.Add("version", "1.0");
            handshakeData.Add("minimumVersion", "0.9");
            handshakeData.Add("supportedConnectionTypes", new List<String> { "long-polling" });
            handshakeData.Add("advice", new Dictionary<String, int>() { { "timeout", 60000 }, { "interval", 0 } });
            List<Dictionary<String, Object>> responseList = Request("/meta/handshake", handshakeData, "handshake");
            clientId = (String)responseList[0]["clientId"];
        }

        public void Connect()
        {
            if(isConnected) {
                return ;
            }
            
            Handshake();

            Dictionary<String, Object> subscribeData = new Dictionary<string, object>();
            subscribeData.Add("subscription", messagingChannel);
            Request("/meta/subscribe", subscribeData);

            isConnected = true;

            new Thread(new ThreadStart(Poll)).Start();

            new Thread(new ThreadStart(PollQueue)).Start();
        }

        public void Disconnect()
        {
            Request("/meta/disconnect", null, "", true);
            isConnected = false;
        }

        private void Poll()
        {
            while (isConnected)
            {
                Request("/meta/connect", null, "connect", false);
            }
        }

        private void PollQueue()
        {
            while (isConnected)
            {
                ProcessMessage(messageQueue.Take());
            }
        }

        private void ProcessMessage(Dictionary<String, Object> data)
        {
            Guid requestId = Guid.Parse((String)data["requestId"]);
            Query query = queries[requestId];

            query.OnMessage(data);
            if (query.isFinished)
            {
                queries.Remove(requestId);
            }
        }

        public void DoQuery(Dictionary<String, Object> query, QueryHandler queryHandler)
        {
            Guid requestId = Guid.NewGuid();
            queries.Add(requestId, new Query(query, queryHandler));
            query.Add("requestId", requestId);
            Request("/service/query", new Dictionary<String, Object>() { { "data", query } });
        }


    }
}

File: lib/importio-client-libs-master/objective-c/Importio.m
//
//  importio.m
//  iOS Client Library
//
//  Created by dev on 23/09/2013.
//  Copyright (c) 2013 importio. All rights reserved.
//

#import "Importio.h"
#import "Queue.h"

@implementation Importio

// message ID
int msgId;

bool _connected;

NSString* clientId;

NSString* _userId;
NSString* _apiKey;

NSString* rootUrl = @"https://query.import.io";

Queue* messageQueue;
NSString* messagingChannel = @"/messaging";

NSMutableDictionary* queries;

- (id) init {
    messageQueue = [[Queue alloc] init];
    queries = [[NSMutableDictionary alloc] init];
    
    _connected = FALSE;
    
    return self;
}

- (id) initWithUserIdAndApikey:(NSString *)userId withApikey:(NSString *)apiKey
{
    _userId = [userId copy];
    _apiKey = [apiKey copy];
    
    messageQueue = [[Queue alloc] init];
    queries = [[NSMutableDictionary alloc] init];
    
    _connected = FALSE;
    
    return self;
}

- (NSString*)urlEncode: (NSString*) unescaped
{
    return (NSString *)CFBridgingRelease(CFURLCreateStringByAddingPercentEscapes(NULL, (__bridge CFStringRef)unescaped, NULL, CFSTR("!*'();:@&=+$,/?%#[]"), kCFStringEncodingUTF8));
}

- (void)login:(NSString *) username withPassword:(NSString *) password withHost:(NSString *)host
{
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
    
    NSString *post = [NSString  stringWithFormat:@"username=%@&password=%@",[self urlEncode:username],[self urlEncode:password]];
    
    NSData *postData = [post dataUsingEncoding:NSUTF8StringEncoding allowLossyConversion:NO];
    
    NSString *postLength = [NSString stringWithFormat:@"%d", [postData length]];
    
    [request setURL:[NSURL URLWithString:[host stringByAppendingString:@"/auth/login"] ]];
    [request setHTTPMethod:@"POST"];
    [request setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Content-Type"];
    [request setValue:postLength forHTTPHeaderField:@"Content-Length"];
    [request setHTTPBody:postData];
    
    NSHTTPURLResponse *response;
    NSData *returnData = [NSURLConnection sendSynchronousRequest:request returningResponse:&response error:nil];
    NSLog(@"%@",[NSString stringWithFormat:@"status:%d", [response statusCode]]);
    NSLog(@"%@",[[NSString alloc] initWithData:returnData encoding:NSUTF8StringEncoding]);
    
    
}

- (NSArray*) request: (NSString*) channel withPath: (NSString*) path withData: (NSMutableDictionary*) data withThrows: (BOOL) throw
{
    
    // Add in the column values
    data[@"channel"] = channel;
    data[@"connectionType"] = @"long-polling";
    data[@"id"] =  [NSString stringWithFormat:@"%d", msgId];
    
    msgId++;
    
    if (clientId != NULL)
    {
        data[@"clientId"] = clientId;
    }
    
    NSString* url = [NSString stringWithFormat:@"%@/query/comet/%@", rootUrl, path];
    
    if (_apiKey != NULL)
    {
        url = [NSString stringWithFormat:@"%@?&_user=%@&_apikey=%@", url,[self urlEncode:_userId],[self urlEncode:_apiKey]];
    }
    
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
    
    [request setHTTPShouldHandleCookies:YES];
    
    NSHTTPCookieStorage* cookieJar = [NSHTTPCookieStorage sharedHTTPCookieStorage];
    
    NSDictionary* headers = [NSHTTPCookie requestHeaderFieldsWithCookies:[cookieJar cookies]];
    [request setAllHTTPHeaderFields:headers];
    
    NSData* postData;
    
    if (data != nil) {
        postData = [NSJSONSerialization dataWithJSONObject:[NSArray arrayWithObjects: data, nil]
                                                   options:kNilOptions
                                                     error:nil];
    }
    
    NSString* postLength;
    
    if (postData != nil)
    {
        postLength = [NSString stringWithFormat:@"%d", [postData length]];
        [request setValue:postLength forHTTPHeaderField:@"Content-Length"];
        [request setHTTPBody:postData];
    }
    
    [request setURL:[NSURL URLWithString:url ]];
    [request setHTTPMethod:@"POST"];
    [request setValue:@"application/json;charset=UTF-8" forHTTPHeaderField:@"Content-Type"];
    [request setValue:@"gzip" forHTTPHeaderField:@"Accept-Encoding"];
    
    NSHTTPURLResponse *response;
    NSError* error = nil;
    NSData *returnData = [NSURLConnection sendSynchronousRequest:request returningResponse:&response error:&error];
    
    if (error)
    {
        NSLog(@"Error occurred %@",error);
    }
    
    NSArray* jsonMessages = [NSJSONSerialization JSONObjectWithData:returnData
                                                            options:kNilOptions
                                                              error:nil];
    
    for(NSDictionary* message in jsonMessages)
    {
        if ([message objectForKey:@"successful"] != nil && [message[@"successful"]isEqual:@"false"])
        {
            NSString* errorMsg = [[NSString alloc] initWithData:returnData encoding:NSUTF8StringEncoding];
            NSLog(@"%@", [NSString stringWithFormat:@"Unsuccessful request: %@", errorMsg]);
            if(throw)
            {
                @throw [NSException exceptionWithName:@"Import io Exception" reason:@"request was not successfully sent" userInfo:nil];
            }
            
        }
        if ([message objectForKey:@"data"]) {
            [messageQueue enqueue:message[@"data"]];
        }
        
    }
    return jsonMessages;
}

- (void) handshake
{
    NSMutableDictionary* handshakeData = [@{@"version":@"1.0",@"minimumVersion":@"0.9",@"supportedConnectionTypes":@[@"long-polling"],@"advice":@{@"timeout":@60000,@"interval":@0}} mutableCopy];
    NSArray* handshake = [self request:@"/meta/handshake" withPath:@"handshake" withData:handshakeData withThrows:TRUE];
    
    // Store a reference to the client id
    clientId = handshake[0][@"clientId"];
}

- (void) connect
{
    if(_connected) {
        return;
    }
    [self handshake];
    
    NSMutableDictionary* subscriptionData = [@{@"subscription":messagingChannel} mutableCopy];
    [self request:@"/meta/subscribe" withPath:@"" withData:subscriptionData withThrows:FALSE];
    
    _connected = TRUE;
    
    NSThread* pollThread = [[NSThread alloc] initWithTarget:self selector:@selector(poll) object:nil];
    [pollThread start];
    
    NSThread* messageThread = [[NSThread alloc] initWithTarget:self selector:@selector(pollQueue) object:nil];
    [messageThread start];
    
}

- (void) disconnect
{
    [self request:@"/meta/disconnect" withPath:@"" withData:[[NSMutableDictionary alloc] init] withThrows:TRUE];
    _connected = FALSE;
}

- (void) pollQueue
{
    while (_connected) {
        [self processMessage:[messageQueue get]];
    }
}

- (void) poll
{
    while (_connected) {
        [self request:@"/meta/connect" withPath:@"connect" withData:[[NSMutableDictionary alloc] init] withThrows:FALSE];
    }
}

- (void) processMessage: (NSDictionary*) data
{
    @try {
        NSString* reqId = data[@"requestId"];
        Query* query = queries[reqId];
        
        [query onMessage:data];
        
        if ([query finished])
        {
            [queries removeObjectForKey:reqId];
        }
    }
    @catch (NSException* e) {
        NSLog(@"Name:%@", e.name);
        NSLog(@"Reason:%@", e.reason);
    }
}

- (NSString*) uuidString
{
    CFUUIDRef uuid = CFUUIDCreate(kCFAllocatorDefault);
    NSString* uuidStr = (__bridge_transfer NSString*)CFUUIDCreateString(kCFAllocatorDefault, uuid);
    CFRelease(uuid);
    return uuidStr;
    
}

- (void) query: (NSDictionary*) query withCallback:(void(^)(Query*,NSDictionary*)) callback
{
    NSString* requestId = [self uuidString];
    queries[requestId] = [[Query alloc] initWithInputsAndCallback:query withRequestId:requestId withCallback:callback];
    
    NSMutableDictionary* queryWithRequestId = [query mutableCopy];
    queryWithRequestId[@"requestId"] = requestId;
    NSMutableDictionary* queryData = [@{@"data":queryWithRequestId} mutableCopy];
    [self request:@"/service/query" withPath:@"" withData:queryData withThrows:FALSE];
}

@end

@implementation Query

- (id) initWithInputsAndCallback: (NSDictionary*) input withRequestId: (NSString*) requestId withCallback: (void (^)(Query*, NSDictionary*)) callback
{
    // Have to copy these here so it doesnt get garbage collected
    _requestId = [requestId copy];
    _input = [input copy];
    _onMessage = callback;
    
    return self;
}

- (void) onMessage: (NSDictionary*) message
{
    NSString* type = message[@"type"];
    
    if([type isEqual:@"INIT"] || [type isEqual:@"START"]) {
        _jobsStarted++;
    } else if([type isEqual:@"SPAWN"]) {
        _jobsSpawned++;
    } else if([type isEqual:@"STOP"]) {
        _jobsCompleted++;
    }
    
    _finished = _jobsStarted == _jobsCompleted && _jobsSpawned+1 == _jobsStarted && _jobsStarted >0;
    
    // If there is an error or the user is not authorised correctly then allow finished to returne true
    if([type isEqual:@"ERROR"] || [type isEqual:@"UNAUTH"] || [type isEqual:@"CANCEL"]) {
        _finished = true;
    }
    
    _onMessage(self,message);
}

@end
